src/modules/timer/Timer.cpp

```diff
@@ -109,52 +100,92 @@ double Timer::getAverageDelta() const
 	return averageDelta;
 }
 
-double Timer::getTimerPeriod()
+#if defined(LOVE_LINUX)
+
+static inline timespec getTimeOfDay()
 {
-#if defined(LOVE_MACOSX) || defined(LOVE_IOS)
-	mach_timebase_info_data_t info;
-	mach_timebase_info(&info);
-	return (double) info.numer / (double) info.denom / 1000000000.0;
-#elif defined(LOVE_WINDOWS)
-	LARGE_INTEGER temp;
-	if (QueryPerformanceFrequency(&temp) != 0 && temp.QuadPart != 0)
-		return 1.0 / (double) temp.QuadPart;
-#endif
-	return 0;
+	timeval t;
+	gettimeofday(&t, NULL);
+	return timespec { t.tv_sec, t.tv_usec * 1000 };
 }
 
-double Timer::getTime()
+static timespec getTimeAbsolute()
 {
-	// The timer period (reciprocal of the frequency.)
-	static const double timerPeriod = getTimerPeriod();
-
-#if defined(LOVE_LINUX)
-	(void) timerPeriod; // Unused on linux
-
-	double mt;
 	// Check for POSIX timers and monotonic clocks. If not supported, use the gettimeofday fallback.
 #if _POSIX_TIMERS > 0 && defined(_POSIX_MONOTONIC_CLOCK) \
 && (defined(CLOCK_MONOTONIC_RAW) || defined(CLOCK_MONOTONIC))
-	timespec t;
+
 #ifdef CLOCK_MONOTONIC_RAW
 	clockid_t clk_id = CLOCK_MONOTONIC_RAW;
 #else
 	clockid_t clk_id = CLOCK_MONOTONIC;
 #endif
+
+	timespec t;
 	if (clock_gettime(clk_id, &t) == 0)
-		mt = (double) t.tv_sec + (double) t.tv_nsec / 1000000000.0;
+		return t;
 	else
+		return getTimeOfDay();
 #endif
-		mt = getTimeOfDay();
-	return mt;
+	return getTimeOfDay();
+}
+
+double Timer::getTime()
+{
+	static const timespec start = getTimeAbsolute();
+	const timespec now = getTimeAbsolute();
+	const timespec rel = timespec {
+		now.tv_sec - start.tv_sec,
+		now.tv_nsec - start.tv_nsec
```

Comment:
I think this math won't work in all situations, eg a start time of 5.8 and a "now" time of 6.1.

I found this page after a quick google which has a fairly complicated bit of code at the bottom for subtracting *spec values: http://elias.rhi.hi.is/libc/Elapsed-Time.html . Maybe something cleaner exists elsewhere though.

Reply:
When I wrote this I just checked the POSIX docs for `timespec` and they said that `timespec.tv_nsec` is a `long` (i.e. signed). Same with the C11 standard, so I thought it was fine. And on POSIX `tv_sec` (`time_t`) should be signed too. It is true however that rel is not a valid timespec object, because `tv_nsec` might be negative. I should probably add a comment that we don't care about rel being a somewhat malformed timespec or rather use two variables with specifically signed types instead.

I think that code you linked is concerned with turning it into a valid timespec. Or am I misunderstanding and it also doesn't work if those fields are both signed?

Reply:
Oh, I didn't realize they were signed. Looks good in that case (unless I'm missing something myself with fractional math).